File: api/app/ingest/stages/chunk_filter.py
"""Chunk 자동 필터링 룰 — G(3) 휴리스틱 마킹 (W3 v0.5 §3.G + W4-Q-15).

배경 (work-log/2026-04-29 청킹 정책 검토.md)
- chunk.py 본격 변경은 W4-Q-14 로 deferred. W3 안에서는 (1) 마이그레이션 004 의 flags
  컬럼 + (2) 본 모듈의 자동 마킹 + (3) backfill 스크립트로 가시성·검색 품질만 확보.
- diagnose_chunk_quality (G(1)) 의 휴리스틱 정의를 그대로 사용 — 임계값만 강화.

휴리스틱 (G(1) 와 정합)
- table_noise: 짧은 라인 ≥ _SHORT_LINE_RATIO_TH AND 숫자/특수문자 ≥ _DIGIT_PUNCT_RATIO_TH
- header_footer: 같은 doc 안에서 동일 텍스트 ≥ _HEADER_FOOTER_REPEAT_TH 회 + len < _HEADER_FOOTER_MAX_LEN
- empty: text.strip() == "" — W4-Q-15 (c) 추가, dead branch 보호 (chunk.py 가 빈 단락 skip 하지만
  향후 DOCX/PPTX 등 새 파서에서 빈 청크 발생 가능성 대비 인프라)
- extreme_short: 1 ≤ len(text.strip()) < _EXTREME_SHORT_LEN — W6 Day 3 추가, 표 셀 격리 부작용
  (단일 숫자/짧은 토큰 청크가 검색 ranking 차지) 회수. DE-65 본 적용 후 G-015 fail 분석 결과.

임계값 강화 (G(1) 대비 — 사용자 결정 2026-04-29)
- short_line_ratio: 0.70 → 0.90 (G(1) 진단용은 0.70, 본 자동 마킹은 0.90)
- digit_punct_ratio: 0.50 → 0.70 (마찬가지)
- 사유: 자동 마킹은 검색에서 제외되므로 false positive 비용이 진단보다 큼. 보수적으로.

마킹 결과
- chunk.flags["filtered_reason"] = "table_noise" | "header_footer" | "empty"
- search_hybrid_rrf RPC 의 WHERE `flags->>'filtered_reason' IS NULL` 가 자동 제외.

가시성 (W4-Q-15 G-2 보강)
- stage 로그에 마킹 카테고리별 카운트 + 비율 노출 (W3 Day 4 ship)
- W4-Q-15 추가: empty 카운트 노출 + 전체 마킹 비율 5% 초과 시 WARNING (false positive risk
  early signal)

위치 (파이프라인 내)
- chunk → **chunk_filter** → content_gate → tag_summarize → ...
- content_gate 가 metadata.pii_ranges / watermark_hits 만 추가 (flags 미터치) 라
  순서 정합 — flags 마킹은 보존됨.
"""

from __future__ import annotations

import dataclasses
import logging
import re
from collections import Counter

from app.adapters.vectorstore import ChunkRecord
from app.ingest.jobs import stage

logger = logging.getLogger(__name__)

_STAGE = "chunk_filter"

# ---------------------- 휴리스틱 임계값 (자동 마킹용 — G(1) 보다 강화) ----------------------

# 짧은 라인 기준 (chars). G(1) 진단과 동일 — 표 셀 통상 길이.
_SHORT_LINE_LEN = 30
# table_noise 자동 마킹 — G(1) (0.70 / 0.50) 보다 보수적 (false positive 비용 ↑).
_SHORT_LINE_RATIO_TH = 0.90
_DIGIT_PUNCT_RATIO_TH = 0.70

# header_footer 자동 마킹 — G(1) 진단과 동일 (이쪽은 보수적이라 강화 불필요).
_HEADER_FOOTER_REPEAT_TH = 3
_HEADER_FOOTER_MAX_LEN = 100

# W6 Day 3 — 표 셀 격리 부작용 청크 자동 마킹.
# 한국어/영문 알파벳 0 + 길이 < _EXTREME_SHORT_LEN → 검색 결과로 의미 0 ("2", "2,800" 등).
# 한국어/영문 1자라도 있으면 의미 있을 가능성 보존 ("변제충당" 4자 보존, "2,800원" 5자 보존).
_EXTREME_SHORT_LEN = 20

# 한국어/영문 글자 vs 숫자/특수문자 분류 (G(1) 동일 정규식).
_DIGIT_PUNCT_PATTERN = re.compile(r"[\d\W_]", re.UNICODE)


def run_chunk_filter_stage(
    job_id: str, *, doc_id: str, chunks: list[ChunkRecord]
) -> list[ChunkRecord]:
    """chunks 에 자동 필터링 룰 적용 — flags["filtered_reason"] 마킹 후 반환.

    부수효과: stage 로그만 (DB chunks 갱신은 load 스테이지에서). 호출 순서는
    pipeline.py 가 책임 — chunk 직후, content_gate 직전.
    """
    with stage(job_id, _STAGE):
        if not chunks:
            return chunks

        # 1) header_footer 후보 — 같은 doc 안에서 반복 짧은 텍스트.
        header_footer_texts = _detect_header_footer_texts(chunks)

        # 2) chunk 별 마킹.
        filtered_count: Counter[str] = Counter()
        out: list[ChunkRecord] = []
        for chunk in chunks:
            reason = _classify_chunk(chunk, header_footer_texts)
            if reason is None:
                out.append(chunk)
                continue
            new_flags = dict(chunk.flags)
            # 기존 flags 보존 (W4+ 다른 마커 추가 가능성 대비).
            new_flags["filtered_reason"] = reason
            out.append(dataclasses.replace(chunk, flags=new_flags))
            filtered_count[reason] += 1

        total = len(chunks)
        filter_ratio = (sum(filtered_count.values()) / total) if total else 0.0
        logger.info(
            "chunk_filter: doc=%s total=%d table_noise=%d header_footer=%d "
            "empty=%d extreme_short=%d filter_ratio=%.3f",
            doc_id,
            total,
            filtered_count["table_noise"],
            filtered_count["header_footer"],
            filtered_count["empty"],
            filtered_count["extreme_short"],
            filter_ratio,
        )
        # W4-Q-15 G-2 보강 — 5% 초과 시 false positive risk early signal
        if filter_ratio > 0.05:
            logger.warning(
                "chunk_filter: doc=%s 마킹 비율 %.1f%% > 5%% — false positive risk 검토 필요",
                doc_id,
                filter_ratio * 100,
            )
        return out


# ---------------------- 휴리스틱 ----------------------


def _classify_chunk(
    chunk: ChunkRecord, header_footer_texts: set[str]
) -> str | None:
    """단일 chunk 의 필터링 사유 — None 이면 통과."""
    text = chunk.text or ""
    stripped = text.strip()

    # W4-Q-15 (c) — 빈 청크 마킹. chunk.py 가 빈 단락 skip 해 dead branch 이지만
    # DOCX/PPTX 등 신규 파서 도입 시 빈 청크 발생 risk 대비 인프라.
    if not stripped:
        return "empty"

    # W6 Day 3 — 표 셀 격리 부작용 회수. 한국어/영문 알파벳 0 + 짧은 청크 (단일 숫자/통화·표 셀 등)
    # 이 검색 ranking 차지하는 케이스 (DE-65 본 적용 후 G-015 fail 분석) 차단.
    if len(stripped) < _EXTREME_SHORT_LEN and not _has_meaningful_letter(stripped):
        return "extreme_short"

    # header_footer 우선 — 짧은 반복 텍스트는 표보다 헤더/푸터 의도가 강함.
    if stripped in header_footer_texts:
        return "header_footer"

    # table_noise — 길이 0 / 너무 짧은 청크는 분류 의미 없음 (다른 단계에서 처리).
    if len(text) < 50:
        return None

    short_line_ratio, digit_punct_ratio = _line_metrics(text)
    if (
        short_line_ratio >= _SHORT_LINE_RATIO_TH
        and digit_punct_ratio >= _DIGIT_PUNCT_RATIO_TH
    ):
        return "table_noise"

    return None


def _has_meaningful_letter(text: str) -> bool:
    """한국어 음절 또는 영문 알파벳 1자라도 포함하면 True (의미 있는 단어 가능성)."""
    return any(c.isalpha() or "\uAC00" <= c <= "\uD7A3" for c in text)


def _line_metrics(text: str) -> tuple[float, float]:
    """(short_line_ratio, digit_punct_ratio) — G(1) 의 compute_chunk_metrics 와 동일 정의."""
    lines = text.split("\n")
    line_count = len(lines)
    if line_count == 0:
        return 0.0, 0.0
    short_lines = sum(1 for ln in lines if len(ln.strip()) < _SHORT_LINE_LEN)
    short_line_ratio = short_lines / line_count

    digit_punct_count = sum(
        1 for ch in text if _DIGIT_PUNCT_PATTERN.match(ch)
    )
    non_ws_total = sum(1 for ch in text if not ch.isspace())
    digit_punct_ratio = (
        digit_punct_count / non_ws_total if non_ws_total else 0.0
    )

    return short_line_ratio, digit_punct_ratio


def _detect_header_footer_texts(chunks: list[ChunkRecord]) -> set[str]:
    """같은 doc 안에서 동일 짧은 텍스트가 임계값 이상 반복되면 헤더/푸터 의심.

    chunks 는 한 번에 단일 doc_id 입력 (pipeline 호출 단위). G(1) 의
    detect_header_footer_candidates 와 같은 알고리즘 — 다만 본 단계는 doc 내부만.
    """
    text_counts: Counter[str] = Counter()
    for chunk in chunks:
        t = (chunk.text or "").strip()
        if t and len(t) < _HEADER_FOOTER_MAX_LEN:
            text_counts[t] += 1
    return {
        t for t, n in text_counts.items() if n >= _HEADER_FOOTER_REPEAT_TH
    }

File: supabase/functions/_shared/ingest/chunk_filter.ts
/**
 * `ingest/stages/chunk_filter.py` 포팅 — 검색에서 뺄 청크를 **마킹**한다(삭제 아니다).
 *
 * `flags.filtered_reason` 이 붙으면 `search_hybrid_rrf` 의
 * `WHERE flags->>'filtered_reason' IS NULL` 이 자동으로 제외한다. 행은 남아 있어서
 * 나중에 판정을 바꾸면 되돌릴 수 있다.
 *
 * | 사유 | 규칙 |
 * |---|---|
 * | `empty` | `strip()` 이 빈 문자열 |
 * | `extreme_short` | 20 자 미만 + **한글/영문 글자가 하나도 없음** (표 셀 "2,800" 같은 것) |
 * | `header_footer` | 같은 문서에서 100 자 미만 텍스트가 3 회 이상 반복 |
 * | `table_noise` | 50 자 이상 + 짧은 줄 비율 ≥ 0.90 + 숫자·기호 비율 ≥ 0.70 |
 *
 * 판정 **순서가 규칙의 일부다** — `header_footer` 가 `table_noise` 보다 먼저다.
 * 짧고 반복되는 텍스트는 표보다 머리말 의도가 강하다는 게 원본의 판단이다.
 *
 * ## 임계값이 진단용보다 높다
 * 같은 휴리스틱을 쓰는 진단 도구는 0.70/0.50 인데 여기는 0.90/0.70 이다. 자동 마킹은
 * 검색에서 빠지므로 오탐 비용이 진단보다 크다 — 보수적으로 잡은 것이다.
 */

import { PY_WORD_CLASS, pyIsSpace } from "../pychar.ts";
import { pyStrip } from "../search/pystr.ts";
import type { ChunkRecord } from "./chunk_records.ts";

/** 짧은 줄 기준(글자 수). 표 셀의 통상 길이다. */
const SHORT_LINE_LEN = 30;
const SHORT_LINE_RATIO_TH = 0.90;
const DIGIT_PUNCT_RATIO_TH = 0.70;

const HEADER_FOOTER_REPEAT_TH = 3;
const HEADER_FOOTER_MAX_LEN = 100;

/** 한글·영문이 0 이고 이보다 짧으면 검색에서 의미가 없다고 본다. */
const EXTREME_SHORT_LEN = 20;

/** `table_noise` 는 이보다 짧은 청크에서는 판정하지 않는다. */
const TABLE_NOISE_MIN_LEN = 50;

/** Python `len()` 은 코드포인트 수다. */
function cpLen(s: string): number {
  let n = 0;
  for (const _ of s) n++;
  return n;
}

/**
 * 원본 `_DIGIT_PUNCT_PATTERN = re.compile(r"[\d\W_]", re.UNICODE)` 의 한 글자 매칭.
 *
 * JS 로 그대로 옮기면 안 된다 — JS 의 `\d` 는 ASCII 숫자뿐이고 `\w` 도 ASCII 다.
 * Python 은 `\d` = 유니코드 `Nd`, `\w` = 글자·숫자·밑줄이다.
 *
 * 뜻을 풀면 **"글자가 아닌 것"** 이다: `\W`(단어문자 아님) ∪ `\d`(숫자) ∪ `_`.
 * 숫자와 밑줄은 `\w` 라 `\W` 에서 빠지는데 앞의 두 항이 도로 넣는다.
 */
const WORD_RE = new RegExp(`^${PY_WORD_CLASS}$`, "v");
const ND_RE = /^\p{Nd}$/u;
function isDigitPunct(ch: string): boolean {
  if (ND_RE.test(ch)) return true;
  if (ch === "_") return true;
  return !WORD_RE.test(ch); // `\W`
}

/** Python `str.isalpha()` — 유니코드 `L*` 범주. */
const ALPHA_RE = /^\p{L}$/u;

/**
 * 원본 `_has_meaningful_letter` — 한글 음절이나 영문 글자가 하나라도 있으면 true.
 *
 * 뒤의 한글 범위 검사는 사실 앞의 `isalpha()` 에 포함된다(한글 음절은 `Lo`).
 * 그래도 원본 그대로 둔다 — 판정이 같고, 원본을 읽는 사람이 대조하기 쉽다.
 */
export function hasMeaningfulLetter(text: string): boolean {
  for (const c of text) {
    if (ALPHA_RE.test(c)) return true;
    if (c >= "가" && c <= "힣") return true;
  }
  return false;
}

/** 원본 `_line_metrics` — (짧은 줄 비율, 숫자·기호 비율). */
export function lineMetrics(text: string): [number, number] {
  const lines = text.split("\n");
  const lineCount = lines.length;
  // 원본에 `if line_count == 0` 가드가 있지만 `"".split("\n")` 은 `[""]` 라 0 이 안 된다.
  // 죽은 가지다 — 옮기되 그 사실을 적어 둔다.
  if (lineCount === 0) return [0.0, 0.0];
  let shortLines = 0;
  for (const ln of lines) if (cpLen(pyStrip(ln)) < SHORT_LINE_LEN) shortLines++;
  const shortLineRatio = shortLines / lineCount;

  let digitPunct = 0;
  let nonWs = 0;
  for (const ch of text) {
    if (isDigitPunct(ch)) digitPunct++;
    if (!pyIsSpace(ch)) nonWs++;  // Python `str.isspace()`
  }
  const digitPunctRatio = nonWs ? digitPunct / nonWs : 0.0;
  return [shortLineRatio, digitPunctRatio];
}

/** 원본 `_detect_header_footer_texts` — 문서 안에서 반복되는 짧은 텍스트. */
export function detectHeaderFooterTexts(chunks: ChunkRecord[]): Set<string> {
  const counts = new Map<string, number>();
  for (const c of chunks) {
    const t = pyStrip(c.text ?? "");
    if (t !== "" && cpLen(t) < HEADER_FOOTER_MAX_LEN) {
      counts.set(t, (counts.get(t) ?? 0) + 1);
    }
  }
  const out = new Set<string>();
  for (const [t, n] of counts) if (n >= HEADER_FOOTER_REPEAT_TH) out.add(t);
  return out;
}

/** 원본 `_classify_chunk` — `null` 이면 통과. **순서가 규칙이다.** */
export function classifyChunk(
  chunk: ChunkRecord,
  headerFooterTexts: Set<string>,
): string | null {
  const text = chunk.text ?? "";
  const stripped = pyStrip(text);

  if (stripped === "") return "empty";
  if (cpLen(stripped) < EXTREME_SHORT_LEN && !hasMeaningfulLetter(stripped)) {
    return "extreme_short";
  }
  if (headerFooterTexts.has(stripped)) return "header_footer";
  // 짧은 청크는 표 노이즈 판정을 하지 않는다 — 다른 가지가 이미 처리했다.
  if (cpLen(text) < TABLE_NOISE_MIN_LEN) return null;

  const [shortLineRatio, digitPunctRatio] = lineMetrics(text);
  if (shortLineRatio >= SHORT_LINE_RATIO_TH && digitPunctRatio >= DIGIT_PUNCT_RATIO_TH) {
    return "table_noise";
  }
  return null;
}

export interface ChunkFilterResult {
  chunks: ChunkRecord[];
  counts: Record<string, number>;
  filterRatio: number;
}

/**
 * 원본 `run_chunk_filter_stage` — 마킹한 새 배열을 돌려준다. 입력은 안 건드린다.
 *
 * DB 갱신은 하지 않는다. `load` 가 `flags` 를 그대로 실어 보낸다.
 */
export function runChunkFilterStage(chunks: ChunkRecord[]): ChunkFilterResult {
  const counts: Record<string, number> = {
    table_noise: 0,
    header_footer: 0,
    empty: 0,
    extreme_short: 0,
  };
  if (chunks.length === 0) return { chunks, counts, filterRatio: 0.0 };

  const headerFooterTexts = detectHeaderFooterTexts(chunks);
  const out: ChunkRecord[] = [];
  let marked = 0;
  for (const chunk of chunks) {
    const reason = classifyChunk(chunk, headerFooterTexts);
    if (reason === null) {
      out.push(chunk);
      continue;
    }
    const flags = { ...((chunk as { flags?: Record<string, unknown> }).flags ?? {}) };
    flags["filtered_reason"] = reason;
    out.push({ ...chunk, flags } as ChunkRecord);
    counts[reason] = (counts[reason] ?? 0) + 1;
    marked++;
  }
  return { chunks: out, counts, filterRatio: marked / chunks.length };
}
